### PR TITLE
GH-100288: Specialize LOAD_ATTR for simple class attributes.

### PR DESCRIPTION
**Pull request by markshannon**
This PR specializes for things like `obj.x` where:
```Py
class C:
    x = 1
obj = C()
obj.x
```
[Stats](https://github.com/faster-cpython/ideas/blob/main/stats/pystats-2023-06-22-faster-cpython-ed977a1.md)
The miss rate for `LOAD_ATTR_NONDESCRIPTOR_WITH_VALUES` is quite poor at 33%
(Ignore the stats for `LOAD_ATTR_NONDESCRIPTOR_LAZY_DICT`, I've removed it)

* Issue: gh-100288

### CODE REVIEW

File: Python/specialize.c [hunk 1]
```diff
@@ -855,10 +855,14 @@ _Py_Specialize_LoadAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name)
...
-                                SPEC_FAIL_ATTR_CLASS_ATTR_SIMPLE :
-                                SPEC_FAIL_ATTR_NOT_MANAGED_DICT);
+            if ((instr->op.arg & 1) == 0) {
+                if (specialize_attr_loadclassattr(owner, instr, name, descr, kind, false)) {
+                    goto success;
+                }
+            }
+            else {
+                SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_ATTR_CLASS_ATTR_SIMPLE);
+            }
```
Comment (brandtbucher) [inline Python/specialize.c:882]: Nit: I find this flow a bit easier to follow:

```suggestion
            if (instr->op.arg & 1) {
                SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_ATTR_CLASS_ATTR_SIMPLE);
            }
            else if (specialize_attr_loadclassattr(owner, instr, name, descr, kind, false)) {
                goto success;
            }
```

File: Python/specialize.c [hunk 2]
```diff
@@ -1082,9 +1087,9 @@ PyObject *descr, DescriptorClassification kind)
             return 0;
         }
         if (dictoffset == 0) {
-            instr->op.code = LOAD_ATTR_METHOD_NO_DICT;
+            instr->op.code = is_method ? LOAD_ATTR_METHOD_NO_DICT : LOAD_ATTR_NONDESCRIPTOR_NO_DICT;
         }
-        else {
+        else if (is_method) {
```
Comment (brandtbucher) [inline Python/specialize.c:1109]: The case where `is_method == false` isn't handled below. It just falls through, writes the caches, and returns success.

File: Python/specialize.c [hunk 3]
```diff
@@ -1082,9 +1087,9 @@ PyObject *descr, DescriptorClassification kind)
...
             PyObject *dict = *(PyObject **) ((char *)owner + dictoffset);
             if (dict) {
                 SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_ATTR_NOT_MANAGED_DICT);
```
Comment (brandtbucher) [inline Python/specialize.c:1112]: Would `LOAD_ATTR_NONDESCRIPTOR_LAZY_DICT` be uncommon or difficult to implement? Just curious.

File: Python/bytecodes.c [hunk 4]
```diff
@@ -2627,6 +2629,35 @@ dummy_func(
...
+            DEOPT_IF(self_cls->tp_version_tag != type_version, LOAD_ATTR);
+            assert(self_cls->tp_flags & Py_TPFLAGS_MANAGED_DICT);
+            PyDictOrValues dorv = *_PyObject_DictOrValuesPointer(self);
+            DEOPT_IF(!_PyDictOrValues_IsValues(dorv), LOAD_ATTR);
+            PyHeapTypeObject *self_heap_type = (PyHeapTypeObject *)self_cls;
+            DEOPT_IF(self_heap_type->ht_cached_keys->dk_version !=
+                     keys_version, LOAD_ATTR);
+            STAT_INC(LOAD_ATTR, hit);
+            assert(descr != NULL);
+            Py_DECREF(self);
```
Comment (brandtbucher) [inline Python/bytecodes.c]:
```suggestion
            DECREF_INPUTS();
```

File: Python/bytecodes.c [hunk 5]
```diff
@@ -2627,6 +2629,35 @@ dummy_func(
...
+        }
+
+        inst(LOAD_ATTR_NONDESCRIPTOR_NO_DICT, (unused/1, type_version/2, unused/2, descr/4, self -- res2 if (oparg & 1), res)) {
+            PyTypeObject *self_cls = Py_TYPE(self);
+            assert(type_version != 0);
+            DEOPT_IF(self_cls->tp_version_tag != type_version, LOAD_ATTR);
+            assert(self_cls->tp_dictoffset == 0);
+            STAT_INC(LOAD_ATTR, hit);
+            assert(descr != NULL);
+            Py_DECREF(self);
```
Comment (brandtbucher) [inline Python/bytecodes.c]:
```suggestion
            DECREF_INPUTS();
```

**Review by brandtbucher — COMMENTED** [review]
A couple notes (and one bug, I think).

Also, your stats seem to indicate that both the success and hit rates for `LOAD_ATTR` are lower with this change. Are we sure it's worth doing?

If so, this seems like quite a bit of extra code and work just to save a branch on the oparg's low bit at the end of the instruction, in my opinion. I would be surprised if the separate instructions were really more performant.

File: Python/specialize.c (hunk 3 above)
Comment (brandtbucher) [reply Python/specialize.c:1112]: Never mind, I see your note about dropping it now (I'm guessing due to the nasty stats).

**Comment by markshannon** [comment]
The stats for `LOAD_ATTR` change from

Kind | Count | Ratio
-- | -- | --
specialization.deferred | 1149226050 | 14.2%
specialization.deopt | 7438533 | 0.1%
hit | 6549029560 | 80.9%
miss | 394372352 | 4.9%

to 

Kind | Count | Ratio
-- | -- | --
specialization.deferred | 997276207 | 11.7%
specialization.deopt | 8561434 | 0.1%
hit | 7089151162 | 83.0%
miss | 453922508 | 5.3%

which is a modest improvement, but the failure stats are now


|Failure kind | Count | Ratio | 
|---|---:|---:|
| has managed dict | 1,088,824 | 78.9% |
| metaclass attribute | 113,508 | 8.2% |
| method | 54,145 | 3.9% |
| shadowed | 36,806 | 2.7% |
| not managed dict | 35,316 | 2.6% |
| mutable class | 19,767 | 1.4% |
| overridden | 7,724 | 0.6% |
| class method obj | 6,782 | 0.5% |
| class attr descriptor | 6,640 | 0.5% |
| non object slot | 6,120 | 0.4% |
| not in keys | 1,680 | 0.1% |
| non overriding descriptor | 1,072 | 0.1% |
| module attr not found | 800 | 0.1% |
| class attr simple | 710 | 0.1% |
| builtin class method | 660 | 0.0% |

Which means that fixing the materialization of `__dict__` can get the specialization success (ignoring misses) rate up to about 97%, which is quite good.

File: Python/specialize.c (hunk 1 above)
Comment (markshannon) [reply Python/specialize.c:882]: To me, it seems more natural to have the success first, then the failure. A matter of taste, I suppose.

**Comment by markshannon** [comment]
> If so, this seems like quite a bit of extra code and work just to save a branch on the oparg's low bit at the end of the instruction.

The `oparg & 1` case is inserted for calls. It seems unlikely that a non-descriptor is being called much, as both Python functions and method-descriptors are descriptors. The stats support this. "class attr simple" is a tiny 0.1% of specialization failures with this change.

**Review by brandtbucher — APPROVED** [review]

**Comment by brandtbucher** [comment]
> The `oparg & 1` case is inserted for calls. It seems unlikely that a non-descriptor is being called much, as both Python functions and method-descriptors are descriptors. The stats support this. "class attr simple" is a tiny 0.1% of specialization failures with this change.

I'm confused. Wouldn't just adding a branch on `(oparg & 1)` at the end of the existing `LOAD_ATTR_METHOD_WITH_VALUES` and `LOAD_ATTR_NONDESCRIPTOR_NO_DICT` instructions have the exact same effect as this PR, but without adding two whole new instructions that are almost identical to the existing ones?

**Comment by markshannon** [comment]
I see what you mean now. I thought you were talking about the specialization, not the resulting instruction.
The specialization paths are distinct, but the resulting instruction is very similar.

I'll try that.

**Comment by markshannon** [comment]
The generated code is noticeably worse for the conditional case.
Comparing the code for the "action" part, the branchless form looks like this:
```C
    Py_DECREF(self);
    res = Py_NewRef(descr);
    stack_pointer[-1] = res;
    next_instr += 9;
```

but with the `oparg` test it looks like this:

```C
    res2 = Py_NewRef(descr);
    if (oparg & 1) {
       res = self;
    }
    else {
        Py_DECREF(self);
    }
    STACK_GROW(((oparg & 1) ? 1 : 0));
    if (oparg & 1) { stack_pointer[-(((oparg & 1) ? 1 : 0))] = res; }
    stack_pointer[-(1 + ((oparg & 1) ? 1 : 0))] = res2;
    next_instr += 9;
```

So, I think the additional instruction is worth it.

**Comment by markshannon** [comment]
Note, the generated code actually has this sequence in it:
```C
    STACK_GROW((0 ? 1 : 0));
    if (0) { stack_pointer[-(1 + (0 ? 1 : 0))] = res2; }
```
but even the worst C compiler will eliminate that code.

**Comment by brandtbucher** [comment]
> So, I think the additional instruction is worth it.

Is the branchy form actually measurably slower? I would imagine that the C compiler would turn the five `(oparg & 1)` branches into just one or two, and that the resulting code would still be about the same (or better than) the cost of bulking out the interpreter.

It's obviously not a huge deal (they're both okay approaches, which is why I approved this PR), but a branch at the end of the existing instruction feels easier to maintain than two new specialized forms.

**Comment by markshannon** [comment]
> Is the branchy form actually measurably slower?

Maybe not now, but it mostly likely will be when JIT compiled.

Is the branchy form otherwise better?
Not really. It may reduce the number of instructions, but it makes the resulting instruction harder to read and optimize.

**Pull request merged**